nrf52840-hal/src/ieee802154.rs

```diff
@@ -0,0 +1,630 @@
+        radio.set_cca(DEFAULT_CCA);
+        radio.set_channel(DEFAULT_CHANNEL);
+        radio.set_sfd(DEFAULT_SFD);
+        radio.set_txpower(DEFAULT_TXPOWER);
+
+        radio
+    }
+
+    /// Changes the radio channel
+    pub fn set_channel(&mut self, channel: Channel) {
+        self.disable();
+
+        // NOTE(unsafe) radio is currently disabled
+        unsafe {
+            self.radio
+                .frequency
+                .write(|w| w.map().clear_bit().frequency().bits(channel as u8))
+        }
+    }
+
+    /// Changes the Clear Channel Assessment method
+    pub fn set_cca(&mut self, cca: Cca) {
+        self.disable();
+
+        match cca {
+            Cca::CarrierSense => self.radio.ccactrl.write(|w| w.ccamode().carrier_mode()),
+        }
+    }
+
+    /// Changes the Start of Frame Delimiter
+    pub fn set_sfd(&mut self, sfd: u8) {
+        // FIXME don't completely turn off the radio; RXIDLE or TXIDLE are probably OK
+        self.disable();
+
+        self.radio.sfd.write(|w| unsafe { w.sfd().bits(sfd) });
+    }
+
+    /// Changes the TX power
+    pub fn set_txpower(&mut self, power: TxPower) {
+        // FIXME don't completely turn off the radio; RXIDLE or TXIDLE are probably OK
+        self.disable();
+
+        self.radio
+            .txpower
+            .write(|w| w.txpower().variant(power._into()));
+    }
+
+    /// Recevies one radio packet and copies its contents into the given `packet` buffer
+    ///
+    /// This methods returns the `Ok` variant if the CRC included the packet was successfully
+    /// validated by the hardware; otherwise it returns the `Err` variant. In either case, `packet`
+    /// will be updated with the received packet's data
+    pub fn recv(&mut self, packet: &mut Packet) -> Result<(), ()> {
+        // NOTE we do NOT check the address of `packet`; see comment in `Packet::new` for details
+
+        // go to the RXIDLE state
+        self.enable_rx();
+
+        // NOTE(unsafe) DMA transfer has not yet started
+        // set up RX buffer
+        unsafe {
+            self.radio
+                .packetptr
+                .write(|w| w.packetptr().bits(packet.buffer.as_mut_ptr() as u32));
+        }
+
+        // start transfer
+        dma_start_fence();
+        self.radio.tasks_start.write(|w| w.tasks_start().set_bit());
+
+        // wait until we have received something
+        self.wait_for_event(Event::End);
+        dma_end_fence();
+
+        if self.radio.crcstatus.read().crcstatus().bit_is_set() {
+            Ok(())
+        } else {
+            Err(())
+        }
+    }
+
+    /// Tries to send the given `packet`
+    ///
+    /// This method performs Clear Channel Assessment (CCA) first and sends the `packet` only if the
+    /// channel is observed to be *clear* (no transmission is currently ongoing), otherwise no
+    /// packet is transmitted and the `Err` variant is returned
+    pub fn try_send(&mut self, packet: &Packet) -> Result<(), ()> {
```

Comment:
It is good to use `CCAIDLE_TXEN` short for this operation to make sure that the transmission takes place right after CCA. With the design of this function, there is unknown software delay between CCAIDLE event and TXEN task.

Reply:
I see it is applied in `send`. Is there any reason here it is handled by software?